numpy/core/src/multiarray/array_coercion.c

```diff
@@ -865,7 +865,8 @@ PyArray_DiscoverDTypeAndShape_Recursive(
         PyObject *obj, int curr_dims, int max_dims, PyArray_Descr**out_descr,
         npy_intp out_shape[NPY_MAXDIMS],
         coercion_cache_obj ***coercion_cache_tail_ptr,
-        PyArray_DTypeMeta *fixed_DType, enum _dtype_discovery_flags *flags)
+        PyArray_DTypeMeta *fixed_DType, enum _dtype_discovery_flags *flags,
+        int do_copy)
```

Comment:
Please add the new parameter to the parameter list above, with permitted values. I do wonder if we could fold it into `_dtype_discovery_flags`, but that can be done in a future refactor if it is complicated or controversial. This is all internal API.